libsignal-service/src/sender.rs

```diff
@@ -70,6 +100,67 @@ where
         }
     }
 
+    /// Encrypts and uploads an attachment
+    ///
+    /// Contents are accepted as an owned, plain text Vec, because encryption happens in-place.
+    pub async fn upload_attachment(
+        &mut self,
+        spec: AttachmentSpec,
+        mut contents: Vec<u8>,
+    ) -> Result<AttachmentPointer, AttachmentUploadError> {
+        let len = contents.len();
+        // Encrypt
+        let (key, iv) = {
+            use rand::RngCore;
+            let mut key = [0u8; 64];
+            let mut iv = [0u8; 16];
+            // thread_rng is guaranteed to be cryptographically secure
+            rand::thread_rng().fill_bytes(&mut key);
+            rand::thread_rng().fill_bytes(&mut iv);
+            (key, iv)
+        };
+        // XXX upstream Java seems to pad this further
```

Comment:
We may want to fix this in this PR

Reply:
Fixed in the last commit. @gferon would you mind taking a second look that I copied that Math thing correctly from Java?